File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/block/BlockProperties.java
package pize.tests.voxelgame.client.block;

import pize.tests.voxelgame.client.block.model.BlockModel;
import pize.tests.voxelgame.client.block.shape.BlockCollide;
import pize.tests.voxelgame.client.block.shape.BlockCursor;
import pize.tests.voxelgame.client.resources.GameResources;
import pize.tests.voxelgame.main.Direction;
import pize.tests.voxelgame.main.audio.BlockSoundPack;
import pize.tests.voxelgame.main.block.BlockData;

import java.util.ArrayList;
import java.util.List;

import static pize.tests.voxelgame.main.chunk.ChunkUtils.MAX_LIGHT_LEVEL;

public abstract class BlockProperties{
    
    private final byte ID;
    private final short defaultState;

    protected boolean solid;
    protected int lightLevel;
    protected int opacity;
    protected boolean translucent;
    protected BlockSoundPack soundPack;
    protected List<BlockState> states;

    protected BlockProperties(int ID){
        this.ID = (byte) ID;
        this.defaultState = BlockData.getData(ID);
        
        solid = false;
        lightLevel = 0;
        opacity = 0;
        translucent = false;
        soundPack = BlockSoundPack.STONE;
        states = new ArrayList<>(1);
    }

    protected abstract void load(GameResources resources);


    public final byte getID(){
        return ID;
    }
    
    public short getDefaultData(){
        return defaultState;
    }
    
    
    /** Возвращает True если это воздух */
    public final boolean isEmpty(){
        return ID == Blocks.AIR.getID();
    }
    
    /** Возвращает True если блок является источником света */
    public final boolean isGlow(){
        return lightLevel != 0;
    }
    
    /** Возвращает True если блок пропускает свет */
    public boolean isLightTranslucent(){
        return opacity != MAX_LIGHT_LEVEL;
    }

    public boolean isLightTransparent(){
        return opacity == 0;
    }
    
    
    /** Возвращает True если блок имеет форму стандартного вокселя
     * (куб, а не любая сложная модель) */
    public boolean isSolid(){
        return solid;
    }
    
    /** Возвращает уровень света блока */
    public int getLightLevel(){
        return lightLevel;
    }
    
    /** Возвращает непрозрачность блока
     * (например: 0 - стекло, 15 - камень) */
    public int getOpacity(){
        return opacity;
    }

    /** Возвращает True если блок полупрозрачный */
    public boolean isTranslucent(){
        return translucent;
    }


    /** Возвращает все возможные состояния блока */
    public List<BlockState> getStates(){
        return states;
    }

    /** Возвращает одно из состояний блока */
    public BlockState getState(int index){
        return states.get(index);
    }

    protected void newState(Direction facing, BlockModel model, BlockCollide collide, BlockCursor cursor){
        states.add(states.size(), new BlockState(facing, model, collide, cursor));
    }


    /** Возвращает набор звуков для блока */
    public BlockSoundPack getSoundPack(){
        return soundPack;
    }
    
    
    @Override
    public boolean equals(Object object){
        if(this == object)
            return true;
        
        if(object == null || getClass() != object.getClass())
            return false;
        
        final BlockProperties blockProperties = (BlockProperties) object;
        return ID == blockProperties.ID;
    }
    
    @Override
    public int hashCode(){
        return ID;
    }
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/block/BlockProperties.kt
package pize.tests.voxelgame.client.block

import pize.math.vecmath.vector.Vec3f.add
import pize.math.vecmath.vector.Vec3i.add
import pize.tests.voxelgame.client.block.model.BlockModel
import pize.tests.voxelgame.client.block.shape.BlockCollide
import pize.tests.voxelgame.client.block.shape.BlockCursor
import pize.tests.voxelgame.client.resources.GameResources
import pize.tests.voxelgame.main.Direction
import pize.tests.voxelgame.main.audio.BlockSoundPack
import pize.tests.voxelgame.main.block.BlockData
import pize.tests.voxelgame.main.chunk.ChunkUtils

abstract class BlockProperties protected constructor(ID: Int) {
    val iD: Byte
    val defaultData: Short

    /** Возвращает True если блок имеет форму стандартного вокселя
     * (куб, а не любая сложная модель)  */
    var isSolid = false
        protected set
    /** Возвращает уровень света блока  */
    var lightLevel = 0
        protected set

    /** Возвращает непрозрачность блока
     * (например: 0 - стекло, 15 - камень)  */
    var opacity = 0
        protected set
    /** Возвращает True если блок полупрозрачный  */
    var isTranslucent = false
        protected set
    /** Возвращает набор звуков для блока  */
    var soundPack: BlockSoundPack
        protected set
    /** Возвращает все возможные состояния блока  */
    var states: List<BlockState>
        protected set

    init {
        iD = ID.toByte()
        defaultData = BlockData.getData(ID)
        soundPack = BlockSoundPack.STONE
        states = ArrayList(1)
    }

    abstract fun load(resources: GameResources?)
    val isEmpty: Boolean
        /** Возвращает True если это воздух  */
        get() = iD == Blocks.AIR.id
    val isGlow: Boolean
        /** Возвращает True если блок является источником света  */
        get() = lightLevel != 0
    val isLightTranslucent: Boolean
        /** Возвращает True если блок пропускает свет  */
        get() = opacity != ChunkUtils.MAX_LIGHT_LEVEL
    val isLightTransparent: Boolean
        get() = opacity == 0

    /** Возвращает одно из состояний блока  */
    fun getState(index: Int): BlockState {
        return states[index]
    }

    protected fun newState(facing: Direction, model: BlockModel, collide: BlockCollide, cursor: BlockCursor) {
        states.add(states.size, BlockState(facing, model, collide, cursor))
    }

    override fun equals(`object`: Any?): Boolean {
        if (this === `object`) return true
        if (`object` == null || javaClass != `object`.javaClass) return false
        val blockProperties = `object` as BlockProperties
        return iD == blockProperties.iD
    }

    override fun hashCode(): Int {
        return iD.toInt()
    }
}
